src/service/resources-impl.js

```diff
@@ -1307,6 +1288,100 @@ export class Resources {
       }
     }
   }
+
+  /**
+   * Calls iterator on each sub-resource
+   * @param {function(!Resource, number)} iterator
+   */
+  setupVisibilityStateMachine_(vsm) {
+    const prerender = {
+      state: VisibilityState.PRERENDER,
+    };
+    const visible = {
+      state: VisibilityState.VISIBLE,
+    };
+    // Viewer has told us we are no longer active
+    const hidden = {
+      state: VisibilityState.HIDDEN,
+    };
+    // Viewer has told us to pause media but don't unload it.
+    const paused = {
+      state: VisibilityState.PAUSED,
+    };
+    // The browser tab is no longer active
+    const inactive = {
+      state: VisibilityState.INACTIVE,
+    };
+
+    const doPass = () => {
+      // If viewport size is 0, the manager will wait for the resize event.
+      const viewportSize = this.viewport_.getSize();
+      if (viewportSize.height > 0 && viewportSize.width > 0) {
+        if (this.hasMutateWork_()) {
+          this.mutateWork_();
+        }
+        this.discoverWork_();
+        let delay = this.work_();
+        if (this.hasMutateWork_()) {
+          // Overflow mutate work.
+          delay = Math.min(delay, MUTATE_DEFER_DELAY_);
+        }
+        if (this.visible_) {
+          log.fine(TAG_, 'next pass:', delay);
+          this.schedulePass(delay);
+          this.updateScrollHeight_();
+        } else {
+          log.fine(TAG_, 'document is not visible: no scheduling');
+        }
+      }
+    };
+    const pause = () => {
+      this.resources_.forEach(r => r.pause());
+    };
+    const unlayout = () => {
+      this.resources_.forEach(r => r.unlayout());
```

Comment:
We need `unlayout` for `VISIBLE` -> `HIDDEN` (browser tab becomes hidden, but we shouldn't stop playback) and `PAUSED` -> `INACTIVE` (where `pauseCallback` has already been called, no need to call again).

> This is because these states are nested like start and resume on Android

I don't understand what that means.